analysis.rms fixes and deprecation removals

orbeckst opened the pull request:
Fixes #1487 and #720

Changes made in this Pull Request:
 - updated tests from RMSD()
 - rotate WHOLE universe, not just mobile #1487 
 - do not superimpose groupselections #720 
 - only allow groupselections with either `weights="mass"` or `weights=None`
 - doc updates/fixes
 - removed deprecated `wass_weighted` keyword
 - removed unused `weights` keyword from RMSF 
 - updated parts of the tests according to the testing style guide
 - added new function `lib.util.get_weights()` to process weights
 - use `get_weights()` in analysis (except encore, which has more specialized requirements)

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

mnmelo commented:
(And I realized I jumped the gun and ignored the WIP label... sorry!)

orbeckst commented:
Thanks for the comments, will attend to when I get to work on it again. 

--
Oliver Beckstein
email: orbeckst@gmail.com


> Am Aug 4, 2017 um 07:59 schrieb mnmelo <notifications@github.com>:
> 
> (And I realized I jumped the gun and ignored the WIP label... sorry!)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

orbeckst commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -528,8 +528,7 @@ def _single_frame(self):
             # R acts to the left & is broadcasted N times.
-            self._ts.positions[:,:] = (self._mobile_coordinates64[:] *
-                                       self._R)
+            self._ts.positions[:, :] = self._ts.positions * self._R
```
yes

orbeckst replied:
The backtics are [standard with numpy style](https://github.com/numpy/numpy/blob/master/doc/example.py). I think @kain88-de also mentioned the inconsistency in the docs with asterisk/backtics but [our doc style guide](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#writing-docstrings) does not really prescribe anything.

orbeckst replied:
Well... I just updated our example with backtics as in numpy. Feel free to start a discussion :-).

orbeckst commented:
I am also going to fix the calculations of the secondary groups. At the moment this uses qcprot but that does an implicit rotational fit. This was not intended: the fit should *only* come from the fitting group and the rmsd of the secondary groups should be a raw `rmsd()`.

orbeckst commented:
@mnmelo please give it a second look when you can spare a moment. Thanks!

orbeckst commented:
I think travis failures are due to #1614.

orbeckst commented:
I rebased against develop to fix any merge clashes and force pushed. 

Also, the old strange failures mentioned in #1614 are not there anymore and the other tests passed.

@mnmelo can you please review the changes when you have a moment (and assuming that the tests pass on the rebased commits)? Thanks!

mnmelo commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -293,49 +313,63 @@ class RMSD(AnalysisBase):
          do a mass-weighted RMSD fit
-    weights : str/array_like (optional)
-         choose weights. If 'str' uses masses as weights
+    weights : {"mass", None} or array_like (optional)
```
Why the bracket notation vs *or* only for the two first possibilities? Why not just use multiple *or*? As in
`weights : "mass" or None or array_like (optional)`

orbeckst replied:
numpy style says to use the braces notation for fixed alternatives (in this case a string... and interpreting `None` liberally), `array_like` is not a fixed value. Writing it in this way seemed to be the clearest way.

mnmelo replied:
Ah, that makes sense. I asked mostly because also here, in the docs of `process_selection`, you explicitly used the *or ... or* notation.

orbeckst commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -167,21 +167,38 @@ def rmsd(a, b, weights=None, center=False, superposition=False):
+    a : array_like
+        coordinates to align to `b`
+    b : array_like
+        coordinates to align (matching `a`)
```
https://github.com/MDAnalysis/mdanalysis/pull/1581#discussion_r132369096

The backtics are [standard with numpy style](https://github.com/numpy/numpy/blob/master/doc/example.py). I think @kain88-de also mentioned the inconsistency in the docs with asterisk/backtics but [our doc style guide](https://github.com/MDAnalysis/mdanalysis/wiki/Style-Guide#writing-docstrings) does not really prescribe anything.

I updated our example with backtics as in numpy. Feel free to start a discussion :-).

orbeckst replied:
Thanks for suggesting clearer docs here, fixed.

orbeckst commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -293,49 +313,63 @@ class RMSD(AnalysisBase):
+        A list of selections as described for `select`. Each selection
         describes additional RMSDs to be computed *after the structures
-        have be superpositioned* according to `select`. The output contains
+        have been superpositioned* according to `select`. The output contains
```
Changed because it reads more nicely but "superposition" is also correct (and we use `superposition` as a keyword in `rmsd()`).

mnmelo replied:
I wasn't against "superposition" as a noun, only as a verb. Anyway, it seems [I was wrong](http://www.thefreedictionary.com/superposition) :P so just write whatever you prefer!

orbeckst replied:
The wonderful thing about the English language is that you can turn *anything* into a verb.

(German is different and Dutch might be, too: anything can be nounified (hehe).)

One of my favorite Calvin and Hobbes strips: http://www.gocomics.com/calvinandhobbes/1993/01/25/ : **Verbing weirds language.**

![Verbing weirds language.](http://assets.amuniversal.com/6f34d860df950131725e005056a9545d)

orbeckst commented:
@mnmelo addressed your comments and found a few other things
- removed deprecated `wass_weighted` keyword
- removed unused `weights` keyword from RMSF (@kain88-de – I hope that's ok with you, you might have added those in anticipation of this functionality becoming eventually available. But I think it's cleaner and more user-friendly to add that back when we actually have the full functionality.)
- updated parts of the tests according to the testing style guide

@utkbansal – please be careful when working on #1610 , quite a few things in the `test_rms.py` changed with this PR, and merging will likely need a bit of manual attention.

kain88-de commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -488,6 +495,15 @@ def __init__(self, atomgroup, reference=None, select='all',
+             len(self.weights) != self.mobile_atoms.n_atoms:
+            raise TypeError("weights (length {0}) must be of same length as "
+                            "the mobile selection ({1})".format(
+                                len(self.weights), self.mobile_atoms.n_atoms))
```
we have weights checks several times in the analysis modules now. Do you mind factoring this into a function and place it in lib so we can use these checks in more places. It's quite useful.

orbeckst replied:
I'll look into it when I get to work on it again.

orbeckst replied:
P.S.: @kain88-de feel free to review my PRs with a "request changes" instead of suggestive comments – I don't mind. If I feel it's too much or not appropriate I will argue but I think this is how it is supposed to work.

orbeckst replied:
@kain88-de I added `lib.util.get_weights()`

orbeckst replied:
Please see https://github.com/MDAnalysis/mdanalysis/pull/1581/files#r133606106

orbeckst commented on package/MDAnalysis/lib/util.py:
```diff
@@ -1267,6 +1267,54 @@ def fixedwidth_bins(delta, xmin, xmax):
+    """
+    if weights == "mass":
+        try:
+            weights = atoms.masses
```
Using `atoms.masses` breaks the idea that *anything in `lib.util` is independent from AtomGroup* etc. However, I didn't really know of a better place to put it. Furthermore, one *could* get around it with duck-typing `atoms` (or just catching the `TypeError`).

Anyway, suggestions/comments welcome.

kain88-de replied:
Yes it's a shame we break the independency of `lib.util` a little bit. It really only does so lightly because we never import a core module. This just assumes the `atoms` object has an attribute weights. That could also be true for other libraries. I don't know of a better place for this function though. We could make a `analysis.util` module?

Otherwise I like your fix. The `AttributeError` is the correct exception to catch and will also work if the atomgroup doesn't have masses.

kain88-de commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -212,14 +222,14 @@ def rotation_matrix(a, b, weights=None):
     Parameters
     ----------
-    a : array-like
+    a : array_like
```
Can you check with other packages I'm pretty sure there should be a normal hyphen.

kain88-de replied:
You are right. It's an underscore

kain88-de commented on package/MDAnalysis/lib/util.py:
```diff
@@ -1267,6 +1267,54 @@ def fixedwidth_bins(delta, xmin, xmax):
+
+    Returns
+    -------
+    weights : array_like or None
```
I would like that this always returns an array. If none is given  the weights should be uniform.

orbeckst replied:
The reason why I kept the special case `None` is so that this can be directly fed to
```
np.average(positions, weights=weights)
```

`None` saves the multiplication of the `weights` with the positions, broadcasting and what not. It's a completely different code path in `np.average()`. Furthermore, it also saves us time in `rotation_matrix()` and friends. It might not make a big difference for a few coordinates but for large systems I would want to avoid *any* array/array broadcast/multiplication.

I think this specific case is sufficiently well motivated to be kept. Or what is your argument for returning a `np.ones(len(atoms))`?

orbeckst replied:
Btw, using `None` is also necessary for the rms.RMSD(..., groupselections)` option, it wouldn't work with a fixed length array. I would need another special case to keep `None` if it were originally provided.

orbeckst replied:
I haven't changed this behavior and wait for your opinion.

kain88-de replied:
The `None` is OK. It gets special treatment later on for performance reasons. I don't like it that much but I also can't think of a better solution.

kain88-de commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -497,79 +570,38 @@ class AlignTraj(AnalysisBase):
+                 prefix='rmsfit_', weights=None,
                  tol_mass=0.1, strict=False, force=True, in_memory=False,
                  **kwargs):
-        """Initialization
```
Could you leave those here please. We semi documented this in the style guide. See https://github.com/MDAnalysis/mdanalysis/issues/919#issuecomment-236558733. I only stated there that the init parameters should be documented in `init` for abstract classes but I think this should be also true for a normal class. It has the advantage that I can easier check if a attribute is missing in the docs.

orbeckst replied:
The problem with documenting parameters in `__init__()` is that this makes it harder to write numpy-doc conforming docs where the parameters come early and then notes and examples come later. If I want to do this then I can only write a small short blurb in the class docs and then have to do *all* notes and examples in `__init__` after the parameters. 

I would prefer to *not* document `__init__` at all and put everything in the class docs because (1) you have all the relevant class docs in one place and (2) even for derived classes without `__init__` you still have the docs in the same place. 

I will undo my changes here and we should discuss how we want to do this.

orbeckst replied:
The last commit has the parameters documented in init.

kain88-de commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -700,8 +723,6 @@ def sequence_alignment(mobile, reference, match_score=2, mismatch_penalty=-1,
     """
-    import Bio.pairwise2
-
     aln = Bio.pairwise2.align.globalms(
```
do we say in the module docs that MDAnalysis overs sequence alignment?

orbeckst replied:
Do you mean "covers" for "overs"?

> do we say in the module docs that MDAnalysis overs sequence alignment?

Not overtly, but the function is documented as a helper and it is necessary for semi-smart superposition. I wouldn't advertise it broadly because that's a side thing for `align`.

kain88-de replied:
Yes I mean `covers`

kain88-de commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -569,6 +628,15 @@ def save(self, filename=None):
 class RMSF(AnalysisBase):
     r"""Calculate RMSF of given atoms across a trajectory.
 
+    .. Note::
```
please use numpy notes. Also here a good idea is to reference alignment and mention that it has the `in_memory` option for smaller trajectories.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -293,15 +303,34 @@ def _fit_to(mobile_coordinates, ref_coordinates, mobile_atoms,
         array of xyz coordinate of reference center of mass
-    weights : numpy array (optional)
+    weights : ndarray (optional)
         Array to be used for weighted rmsd
```
Ok, but this one cannot use "mass" because `rotation_matrix()` is pretty bare-bones and just uses whatever array is passed through. I'd keep it that way for the lower level functions.

mnmelo commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -569,67 +631,126 @@ def save(self, filename=None):
+        average structure as a reference. That requires calculating the average
+        structure first. Because we need to analyze and manipulate the same
+        trajectory multiple times, we are going to load it into memory using
+        the :module:`~MDAnalysis.coordinates.MemoryReader`. (If your trajectory
```
This seems to be where sphinx is erroring out. I think you meant `:class:`.

orbeckst replied:
It was actually the `:module:`, should have been `:mod:`. Fixed in the force-push.

orbeckst commented:
@kain88-de  https://github.com/MDAnalysis/mdanalysis/pull/1581#pullrequestreview-56877397

> I think encore is also using weights somewhere. This should also be fixed there.

It does (in [confdistmatrix](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/encore/confdistmatrix.py#L139), [covariance](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/encore/covariance.py#L227), [similarity](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/encore/similarity.py#L830)) but in all three cases, `get_weights()` is not flexible enough to deal with the special requirements of encore, at least as far as I can tell. Instead of breaking the existing code and shoe-horning it into `get_weights()` compatibility, I'd rather leave it as it is. Once this PR is merged (whose main purpose was really to fix two bugs and not to rewrite all RMS-related code), we can have a second look – for instance, some of the `if` statements such as [covariance.py#L229](https://github.com/MDAnalysis/mdanalysis/blob/0acb79661255cb39228b8734eb614eb8a3c3f990/package/MDAnalysis/analysis/encore/covariance.py#L229) can be simplified to `if weights == "mass"`.

kain88-de commented on package/MDAnalysis/lib/util.py:
```diff
@@ -1267,6 +1267,54 @@ def fixedwidth_bins(delta, xmin, xmax):
+
+    if iterable(weights):
+        if len(weights) != len(atoms):
+            raise TypeError("weights (length {0}) must be of same length as "
```
This should be a value error.

orbeckst replied:
changed in 125bd7d

kain88-de commented on package/MDAnalysis/lib/util.py:
```diff
@@ -1267,6 +1267,54 @@ def fixedwidth_bins(delta, xmin, xmax):
+                            "the atoms ({1})".format(
+                                len(weights), len(atoms)))
+        elif len(np.asarray(weights).shape) != 1:
+            raise TypeError("weights must be a 1D array, not with shape "
```
This also should be a ValueError.

orbeckst replied:
changed in 125bd7d

kain88-de commented:
> It does (in confdistmatrix, covariance, similarity) but in all three cases, get_weights() is not flexible enough to deal with the special requirements of encore, at least as far as I can tell.

OK. Then we can leave this for some other time.

mnmelo commented:
The 3.6 and minimal builds were erroring out. Restarted them.
The 'minimal' build was failing with one of the spurious 'not a zip' errors. But the 3.6 was failing with `lib/util.py` trying an operation on a closed file. Hopefully also spurious and unrelated to the changes here, but let's wait and see.

kain88-de commented:
Thanks @orbeckst
